### ISSUE TITLE
Blocked UI while loading images in notebook

### ISSUE THREAD

**Issue by pwinston** [napari/napari#1320]
## 🐛 Bug

In Juan's "Intro to Neural Nets" napari notebook scrolling through the 10,000 tests cases is slow because some of the layers are computed on the fly. Unlike #845 (multi-scale) or #1300 (large images) in this case the images are all very small and there is minimal IO.

Instead here the slowness is a machine learning (torch) computation. Accessing the image data via `asarray` triggers a computation that takes some time to produce the data. While waiting for the computation to finish the UI is blocked.

## To Reproduce

1. Get the "napari" branch of https://github.com/jni/noise2self
2. Run `jupyter notebook` then run "Intro To Neural Nets" in the notebooks directory.
3. In cell 15 napari is launched showing a time series of 10,000 small images, 3 layers each.
4. Use the slider to scroll through the time series.

Each image takes maybe 500ms to load on a new macbook. This makes the slider extremely awkward to use, and the overall experience is poor.

## Expected behavior

1. The UI should not be blocked. You should be able to move the slider freely at all times.
2. The `clean_test_dask` layer is not computed on the fly. It should be display as soon as possible, before the other layers are ready. 

With both 1 and 2 working the end result would be almost more like a custom application than just an image viewer. You can quickly scroll through the input images to find one you like, then you pause and wait for the result to be computed and displayed.

Specifically the `clean_test_dask` layer could be shown quickly:

<img width="1085" alt="napari-clean" src="https://user-images.githubusercontent.com/4163446/83364008-ebbbc100-a36b-11ea-9f66-d72f1a63cb2d.png">

Then the other 2 layers added when available:

<img width="1091" alt="napari-all" src="https://user-images.githubusercontent.com/4163446/83364011-f6765600-a36b-11ea-8708-5cca48868f2a.png">

## Additional context

We are writing a document that will propose a solution for the three blocked UI issues: this one, #845 and #1300.

**Comment by tlambert03** [comment]
Just curious: since that’s a delayed dask array, could the the delayed function just be a threadworker?

**Comment by pwinston** [comment]
Not sure what you mean exactly @tlambert03. Do you mean could Juan use`@thread_worker` and make this work as desired with no napari changes?

**Comment by tlambert03** [comment]
That’s what I’m wondering (but definitely not sure).  There’s two slow things happening here: 1) the model is training slowly in another thread, and 2) periodically the model is used for prediction in the main thread.  Every time # 2 happens, dask calls compute on whatever chain of delayed functions are required to fetch/calculate the requested indices.  Seems like we need to get that compute in another thread, so I’m just wondering if the same trick as #1 couldn’t be used.

**Comment by jni** [comment]
@tlambert03 it's not entirely trivial, but a pattern like `thread_worker`, maybe even a subclass, would be handy. I've pointed out the lines that need updating here:

https://github.com/napari/napari/issues/1300#issuecomment-636038648

In short, the trivial approach would be something like replacing `image = np.asarray(<current_slice>)` with:

```python
worker = thead_worker(
    np.asarray,
    connections={'returned': self.set_data},
)(<current_slice>)
```

Which I gotta be honest, is *super clean* and I love it. =D

However, the problem with this approach is that it doesn't let me *cancel* that job once I move to the next slice — which I would, repeatedly, because this would not be blocking my slider anymore. We want to ensure that only the last of these calls is actually running at any one time. So there needs to be some queue management on top of the threadworker. ie I think we need to add `start_thread=False` and push the worker onto a queue that has a max length of 1 or 2 — this will depend on the exact engineering of the work queue, but if it's 1 then you never really get a chance to cancel the work before you start it.

**Comment by tlambert03** [comment]
Yeah makes sense.  I’ll play with it, maybe we can use a semaphore?

**Comment by pwinston** [comment]
I'm still not clear if we are trying to make Juan's example work or trying to add async loading to napari in general. Adding async loading in general is what I was planning to do as part of fixing #845, #1300 and this issue. So I don't know @tlambert03 are you taking on the general problem here?

As Juan says I think you need a queue and you need to constantly cull/cancel "stale" requests. Whatever the user is currently viewing defines some working set of chunks. In this case there's no spatial subdivision so the chunks are just the layers of the slice we are currently viewing.

Every `draw()` we draw all the chunks that are in memory ready to go, and we queue any chunks that are still needed. If the view then changes we have a new working set. We cancel any requests where the load hasn't started, so the queue only ever contains fresh/relevant requests.

Two subtle details are how many requests do we load in parallel and do we cancel in-progress already-running requests, or do we only cancel stale requests that haven't started. I think it both cases it might depend on what's going on in the async part, whether it's IO or computations and if it's computation how many cores is it using.

**Comment by tlambert03** [comment]
no, I'm definitely not trying to take on the general problem! :) ... but it just occurs to me that it's _possible_ that we already have at least some of the necessary machinery in the thread_worker pattern.  It's of course not the only way we could do it... we can also do something with asyncio, or any other python concurrency pattern.  But if we can minimize the number of different concurrency APIs going on, that'd be great :)

**Comment by tlambert03** [comment]
I was able to get a basic queue with max_workers running with the `thread_pool` code, so it doesn't seem far off...  Anyway, I _definitely_ don't want you to think we're duplicating work or anything (or that I'm also taking on the whole thing).  I just also am curious about this problem and will probably be thinking about it as well.  (and as mentioned, it feels like the solution for this has a lot of overlapping concepts with the solution for a user-facing thread worker)

**Comment by tlambert03** [comment]
> how many requests do we load in parallel and do we cancel in-progress already-running requests, or do we only cancel stale requests that haven't started.

for what it's worth, if we were to try to use the `thread_worker` pattern and the data-retrieving function was _not_ a generator (and most would not be)... then there's literally no way to stop the in-process worker.  for that we would need to move to a more complex pattern.  we _could_ prevent a unstarted worker from starting if one is already working and another request comes along, but if we capped the number of workers at 1, the final worker (the one we care about) would always have to wait for the last one to finish before starting.  That does seem to be the most complicated detail here, since it's hard to kill in-process threads that aren't "checking in"

**Comment by tlambert03** [comment]
@pwinston, I made a [branch](https://github.com/tlambert03/napari/tree/wip/threaded-set-view-slice) with a partially working example of using thread_worker to prevent UI blocking.  Note, many of our events assume single-threading, so I had to move some events around (breaking everything but the image layer at the moment)... but `examples/multithreading_slow_data.py` in that branch is currently like this (fetching data takes 500ms):

![Untitled](https://user-images.githubusercontent.com/1609449/83424023-d1c8bf80-a3f9-11ea-9ce7-fbc6d0bb2577.gif)

while it prevents intermediate requests, an obvious remaining issue there is that it doesn't take exactly 500 ms to render an image from the moment when the slider stops... it takes `500ms + <time remaining in previous job>`.  so when you slide the slider from 50 -> 80, it immediately starts working on slice 51 ... throws away everything in between and then starts working on 80 when 51 is done.   So the queueing needs some work.  (side note: this might be a good example of needing to use `QThreads` instead of `QRunnables`)

**Comment by pwinston** [comment]
I spoke to @tlambert03 about this. This is a great prototype and the fuctionality is pretty close to what we want. Definitely proves out the concept I think some changes for the real version are:

1. Create a global `ChunkPager` object with a thread pool, instead of having `WorkerBase` just create an "anonymous" pool under the hood on they fly. We'll want to submit requests to the `ChunkPager` from various places, query it, configure it, and generally have a "thing" that is our pager. We might have multiple pagers, like for different layer types, so maybe create a `ChunkManager` with one or more pagers?

2. Add some type of `DataSource` object that `Image` holds. The data source may or may not have the data in memory. If it does it has what's in `Image._data_raw`basically. 

3. Instead of calling `_set_data` after the chunk is loaded, we'll just set the data in the `DataSource` and then trigger a refresh. Here he moved some stuff into `_set_data` that we obviously can't move in there.

4. We talked about how `_set_view_slice` is more or less just a `draw()` or `render()` method at least for `Image`. We call it to draw the layer even if the slice was not changed at all.

5. We have to figure out how many layers do we load at once (size of the global thread pool) and how do we figure out the right order. We want to load the fastest layers first. Lot's of TBD there.

Other than that it's a pretty good prototype of what we want.

**Comment by tlambert03** [comment]
will also just add (as discussed) that this will require a fair amount of refactoring of the event chain during `layer.refresh()`.  Currently... the sequence of events that needs to occur to update the view is spread across multiple methods that each have side-effects and must be called in just the right order... and that only works in a single-threaded design.

**Comment by pwinston** [comment]
Progress with #1354, not pushed yet. We figured out that we need to convert input arrays with `np.array` because using dask from them is slow due to: https://github.com/dask/dask/issues/5913

See below for a newer/better GIF, I removed this one it's dated.

**Comment by jni** [comment]
That is 😍 @pwinston!

**Comment by pwinston** [comment]
Slider is working now at a reasonable rate. Had to switch to processes from threads, and that currently means it doesn't work with Jupyter. Might require something special. This is a stand-alone script:
![napari-slider-working](https://user-images.githubusercontent.com/4163446/87239164-eb3a2f80-c3d9-11ea-9c79-42da00b093ad.gif)

I haven't yet tried to quantify exactly how much Python-bytecode Torch is running, but it's a lot. When we run Torch in a worker thread the main thread slows WAY down, like from 100ms to 600ms. When moved to a worker process the GUI thread is tons faster. I'll create some viz/tests to quantify and illustrate this, but it's pretty clear.

We could have a config option to use threads or processes. We would even make it a per-layer option. But using concurrent.future’s process pool in Jupyter is giving weird errors. Because it's more or less trying to run Jupyter N times....

We might be able to use an explicit small loader process instead to avoid re-running the entire current executable. Although this is not easy because it needs the dependencies necessary to run the task, like Torch in this case. Jupyter makes it tough, because the original process is all of Jupyter.

**Comment by pwinston** [comment]
After spending most of the week staring at fubar'd traces this is one is exactly right. On far left it intiaties 3 loads, two of them are ndarray so we chose to load them synchronously and that’s basically instant. Then we draw those two. 

The third layer is Dask so we chose to run it in the worker process. In this case it’s running Torch and it takes 625ms. When done we draw that layer. We actually draw all three layers but only one is drawing for the first time which is take 15-20ms to load the texture, the other two are instant be use they are already on the card, in the scene. The GUI is fully interruptible the whole time.

<img width="1184" alt="napari-async-process" src="https://user-images.githubusercontent.com/4163446/87239376-7caaa100-c3dc-11ea-9846-f0258f82b6ce.png">

**Comment by jni** [comment]
@pwinston 

> Slider is working now at a reasonable rate. Had to switch to processes from threads, and that currently means it doesn't work with Jupyter. Might require something special. This is a stand-alone script:

Ah, catching up here, after commenting prematurely in pwinston/napari#5. Do you have your standalone script posted somewhere? And why does jupyter not work with process workers?

**Comment by pwinston** [comment]
I posted the script I was using here:
https://github.com/pwinston/napari-pwinston/blob/master/noise2self/noise2self_script.py

if you run it like:

**Comment by pwinston** [comment]
I posted the script I was using here:
https://github.com/pwinston/napari-pwinston/blob/master/noise2self/noise2self_script.py

if you run the https://github.com/pwinston/napari/pull/5 branch with a config file like this:
NAPARI_ASYNC=~/.napari-async-config python noise2self_script.py

where your config uses processes like:
```
{
    "synchronous": false,
    "num_workers": 6,
    "use_processes": true,
    "log_path": "/Users/pbw/dev/napari/chunk.log",
    "force_delay_seconds": 0
}
```

You'll see the "fast" behavior shows in the last picture above, where the two `ndarray` layers pop in very quickly (~100ms) and the ~600ms Torch computation is in another process.

**Comment by pwinston** [comment]
This is the exception I get in Jupyter. I've not tried to investigate yet, but it did not work "out of the box":

```
exception calling callback for <Future at 0x7fe638fb5ac0 state=finished raised PicklingError>
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/multiprocessing/queues.py", line 239, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function <listcomp>.<lambda> at 0x7fe60ac82f70>: attribute lookup <listcomp>.<lambda> on __main__ failed
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/concurrent/futures/_base.py", line 328, in _invoke_callbacks
    callback(self)
  File "/Users/pbw/dev/napari/napari/utils/chunk/_loader.py", line 251, in _done
    request = future.result()
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/multiprocessing/queues.py", line 239, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function <listcomp>.<lambda> at 0x7fe60ac82f70>: attribute lookup <listcomp>.<lambda> on __main__ failed
exception calling callback for <Future at 0x7fe60dda2a90 state=finished raised PicklingError>
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/multiprocessing/queues.py", line 239, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function <listcomp>.<lambda> at 0x7fe609de09d0>: attribute lookup <listcomp>.<lambda> on __main__ failed
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/concurrent/futures/_base.py", line 328, in _invoke_callbacks
    callback(self)
  File "/Users/pbw/dev/napari/napari/utils/chunk/_loader.py", line 251, in _done
    request = future.result()
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/multiprocessing/queues.py", line 239, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function <listcomp>.<lambda> at 0x7fe609de09d0>: attribute lookup <listcomp>.<lambda> on __main__ failed
exception calling callback for <Future at 0x7fe619845a30 state=finished raised BrokenProcessPool>
Traceback (most recent call last):
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/concurrent/futures/_base.py", line 328, in _invoke_callbacks
    callback(self)
  File "/Users/pbw/dev/napari/napari/utils/chunk/_loader.py", line 251, in _done
    request = future.result()
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```

**Comment by pwinston** [comment]
I am a bit worried about using `concurrent.futures` processes with a big 3rd party program like Jupyter. Instead of creating a "second copy of napari" we are creating "a second copy of jupyter".

We're not actually running a second copy. But the new process as I understand it will import all the same stuff. It's just instead of calling `main()` it will call our worker. But anything "created on import" will exist in every process. So it just seems like that will lead to problems. But above is the only current concrete problem.

The major alternative to concurrent.futures I can think of is make the user define a "worker process" module something like `my_torch_processor.py` in this case. And napari can start up that file along as the worker process, instead of starting Jupyter. But of course it ruins the idea that your worker can be defined in the notebook itself. You'd have to make this separate little "worker process" file and reference it from your notebook. Still it would be "running inside jupyter" in a way.

**Comment by pwinston** [comment]
I think the above error alludes the big difference between threads and processes. All the memory in the process is available to all the threads. So when you pass data "to" worker thread or "from" a worker thread, nothing is passed except references. All the data stays put, zero copies, and is available to everyone.

With processes by default it serializes the data and transfers it somehow between them. By default that means pickle. Early on I hit a problem that it couldn't pickle my weakref Layer objects. Couldn't pickle Layer objects themselves either I'm sure. Because the Layer doesn't exist in the other process. That was easy to fix I added `ChunkLoader.layer_map` and in the `ChunkRequest` I only store `id(layer)` which you can throw around between processes just fine.

The Dask docs suggest pickling large numpy arrays does not work well, which could be an issue! The Dask docs say a lot about serialization and pickingly and they have their own protocols and things.

I'm wondering if shared memory has a role. Dask discusses it but it's not clear which problems it solves and which remain. You maybe still need serialization because the share memory needs to be purely "data" it can't have references to anything because it would not exist in both processes? There is a learning curve here.
https://docs.dask.org/en/latest/shared.html

At any rate we are just scratching the surface with worker processes and we either figure out how to make them work well or we don't.

**Comment by pwinston** [comment]
The other angle here is maybe when we go from threads to processes we should also go from "napari" to "dask". Either regular Dask or Dask.distributed with the single-machine scheduler.

That is, if processes don't work out of the box, and we start down the road of making cross-process stuff work in all cases, that might be redoing what Dask already does better. Although whether you can Dask+napari in your Jupyter notebook I'm not sure.

**Comment by jni** [comment]
Thanks for all this @pwinston! Super tricky issue. For reference, there is a whole thread about it in Zulip here:

https://napari.zulipchat.com/#narrow/stream/212875-general/topic/1320.20jupyter.20script

It's unclear why PyTorch would need processes to work because searching online I can only find mentions saying PyTorch does *not* hold the GIL. I do wonder if there is something else going on here because trying to use async for an IO-bound volume is *also* slow. Using the pwinston#5 branch,

```
NAPARI_ASYNC=1 napari https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr
```

and

```
napari https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr
```

have approximately the same performance. (And only render a black canvas, which is a separate issue.) In those cases, the limiting factor is IO over http, so I *presume* the GIL is not an issue there...

**Comment by sofroniewn** [comment]
Whoo!!! I was able to get it working. I'm trying a 3 camera hand tracking dataset from a collaborator that was super slow to work with before. It's a bit awkward with the screen going black when I scrub around, but the experience if much nicer than before when things would just freeze up and crash with the spinning wheel, and we can figure out something else to do there in the future.

I can share code/ movies for this one if people like, but basically I'm using ffmpeg via pyav via pims via imageio via dask to lazily load each frame of three different `.mov` file. I have the "hand" vectors loaded into memory so they are fast to display

I'm going to play with some more examples, but wanted to share this one now!

![async_multicamera](https://user-images.githubusercontent.com/6531703/88130042-6c3db780-cb8e-11ea-98e4-a0e4b2675486.gif)

Edit: Code link here https://github.com/sofroniewn/image-demos/blob/master/examples/hand_demo-2D.py, I couldn't find public data link but I can share.

I also wanted to try with some multiscale data (even though I know this PR isn't for multiscale data) and ran into this too

> And only render a black canvas, which is a separate issue.

Appologies if this has been stated elsewhere, but I get a black canvas simply from running
`python examples/add_multiscale_image.py` with `NAPARI_ASYNC=0` and `NAPARI_ASYNC=1`, so looks like something isn't quite right for multiscale right now.

**Comment by sofroniewn** [comment]
This also work well for my remote zarr lazy blur/ deconv dataset where there is one small remote dataset and then i lazily perform a blur and a deconvolve using dask delayed. Note here how the raw (bottom-left) always arrives first, then the blurred (top-right) arrives next quite soon after, bluring is fast, but the deconvolved (top-left) takes longer to show up. Pre-async this example is basically un-usable for a user experience perspective

![async_remote_smFISH2](https://user-images.githubusercontent.com/6531703/88133778-bb3c1a80-cb97-11ea-87cc-49e78bce7d82.gif)

I'll note that on initial loading the thumbnails don't appear to be there even though the data is there. They start appearing part way through the demo.

EDIT: here is code https://github.com/sofroniewn/image-demos/blob/master/examples/remote_smFISH.py. Data is public, so hopefully anyone can reproduce.

**Comment by sofroniewn** [comment]
It's also working well for my remote lattice lightsheet data set, just loading each slice as I scroll through "Z" and "Time". It seems to be slow every so often, but not sure if that's just my WIFI or zarr things or something else, but again - so much nice to be slow to load but I'm still free to keep interacting, slider keeps moving etc:

![async_remote_AOLLSM](https://user-images.githubusercontent.com/6531703/88134388-55e92900-cb99-11ea-834e-698a50424b9b.gif)

Code for this one is here, data is public so anyone should be able to demo https://github.com/sofroniewn/image-demos/blob/master/examples/remote_AOLLSM.py

**Comment by jni** [comment]
@sofroniewn to be clear, are you doing:

```
NAPARI_ASYNC=1 python path/to/remote_AOLLSM.py
```

Or...?

**Comment by sofroniewn** [comment]
I ran
```
export NAPARI_ASYNC=1
python path/to/remote_AOLLSM.py
```

**Comment by pwinston** [comment]
Above @jni notes that this is not faster when you turn async on:
```
NAPARI_ASYNC=1 napari https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr
```
The reason is napari thinks the image is multi-scale, and multi-scale is not async yet. I added a log message, when this `Image` is created it logs:
```
Image.__init__ multiscale=1 init_shape=(1, 236, 275, 271)
```
It'd be nice if there were a clearer indication in the UI whether a layer was single-scale or multi-scale. Either in the layer widget or with some type of "layer properties" or "layer status" dialog box that lists the layers and has columns of information about them. For example dimensions, type (rgb, rgba), loading speed, etc. Kind of like Mac's Activity Monitor but showing the layers instead of processes.

**Comment by jni** [comment]
Ah, apologies for the noise @pwinston! There *is* a small indication that it's multiscale in that the reported value under the cursors is "level, value" instead of just "value", but it didn't even occur to me that it would be multiscale, since it's so tiny. I agree that a different icon, maybe a little pyramid made from the picture icon, (ditto for multiscale labels).

**Comment by pwinston** [comment]
Cool as long as it really is multiscale, i.e. the branch is not mistakenly thinking it's multiscale.

**Comment by pwinston** [comment]
I fixed the "multi-scale renders as black" problem in #1495. Multi-scale images were never being marked as "loaded". When an image is not "loaded" the vispy node is invisible.

Long term it might be nice to leave up the previous image until the new one is ready to draw, to avoid going to black. Figuring the "wrong" image at least is a point of reference. For example with the "hand tracking" one above, seeing the hand skeleton on top of the "wrong" image seems like it'd look better than black.

But you do need some visual indication it's the wrong/stale image, and some indication when it's updated to the correct image, when the layers are synced up again. Also it might not be totally simple to implement since you need to keep the old/wrong image around where today we get rid of it as soon as we start loading the new image.

We might have to try a few things until we get something that looks/feels right. We can try to do something now or keep it as black until we are further along.

**Comment by pwinston** [comment]
For above `remove_smFISH.py` @sofroniewn I cannot access the data. It said first need credentials but then I setup my own AWS ones, then it says `Forbidden`. Is the data somewhere else?

```
Traceback (most recent call last):
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/site-packages/s3fs/core.py", line 529, in info
    out = self._call_s3(self.s3.head_object, kwargs, Bucket=bucket,
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/site-packages/s3fs/core.py", line 200, in _call_s3
    return method(**additional_kwargs)
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/site-packages/botocore/client.py", line 316, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/pbw/miniconda3/envs/new-base/lib/python3.8/site-packages/botocore/client.py", line 635, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (403) when calling the HeadObject operation: Forbidden
```

**Comment by sofroniewn** [comment]
I’m not at my laptop now, but I wonder if @tlambert03 was able to access that data. I think getting all this stuff nicely available will be a priority!

**Comment by tlambert03** [comment]
@sofroniewn, I'm not 100% sure, but I think you might still need to adjust the [Bucket Policy](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/add-bucket-policy.html)?

@pwinston, try this dataset to see if it behaves differently  (first make sure to `pip install s3fs`)

```python
import napari
import dask.array as da
data = da.from_zarr("s3://tjl-data/neuro.zarr")
napari.view_image(data, channel_axis=0, contrast_limits=(0,1200))
```

**Comment by tlambert03** [comment]
> but I wonder if @tlambert03 was able to access that data.

I get the Forbidden error as well for  `s3://sofroniewn/image-data/smFISH/raw.zarr`

**Comment by sofroniewn** [comment]
Given the merge of the first wave of async PRs which enable this functionality. I am going to close this issue.
